Mathlib/LinearAlgebra/Dual.lean

```diff
@@ -1231,6 +1231,15 @@ theorem finrank_add_finrank_dualCoannihilator_eq (W : Subspace K (Module.Dual K
   rw [eq.symm, add_comm, Submodule.finrank_quotient_add_finrank, Subspace.dual_finrank_eq]
 #align subspace.finrank_add_finrank_dual_coannihilator_eq Subspace.finrank_add_finrank_dualCoannihilator_eq
 
+theorem _root_.LinearMap.flip_surjective_of_injective {B : V →ₗ[K] Module.Dual K V₁}
+    (hB : Function.Injective B) : Function.Surjective B.flip := by
+  rw [← range_eq_top]
+  apply eq_top_of_finrank_eq
+  rw [dual_finrank_eq, eq_comm, ← finrank_add_finrank_dualCoannihilator_eq,
+      add_right_eq_self, finrank_eq_zero, eq_bot_iff]
+  refine fun x hx ↦ hB (ext fun v ↦ ?_)
+  simpa using (mem_dualCoannihilator x).mp hx _ (mem_range_self _ v)
```

Comment:
You can generalize this to these:
```lean
theorem injective_flip_iff [FiniteDimensional K V₁] {B : V₁ →ₗ[K] V₂ →ₗ[K] K} :
    Function.Injective B.flip ↔ Function.Surjective B := by
  rw [← dualMap_surjective_iff]
  rw [← Equiv.surjective_comp (Module.evalEquiv K V₁).toEquiv]
  rfl

theorem surjective_flip_iff [FiniteDimensional K V₂] {B : V₁ →ₗ[K] V₂ →ₗ[K] K} :
    Function.Surjective B.flip ↔ Function.Injective B := by
  rw [← injective_flip_iff]
  rfl
```
They can go right after `LinearMap.dualMap_surjective_iff`.

I think it's nicer stating them as bilinear forms directly rather than using `Module.Dual`, since `flip` unfolding the definition is a little obscure.

(Do you think these can be generalized further?)

Reply:
Unfortunately, you're proving different results than what I want; I'd like to have [FiniteDimensional K V₁]` instead (V1 rather than V2) in
```
theorem surjective_flip_iff [FiniteDimensional K V₂] {B : V₁ →ₗ[K] V₂ →ₗ[K] K} :
```
Your version is more restrictive, because if V2 is f.d. and B is injective, then V1 is also f.d., while my version allows the possibility of V2 being infinite dimensional. I can't think of a shorter proof of my version at this moment ...

I think your results are still interesting and I'll include them. To distinguish with my version (upgraded to iff), I will use subscripts 1 and 2 to indicate which space needs to be finite dimensional. I'll use the name `flip_surjective_iff` following the convention of `dualMap_surjective_iff` etc. 

I think the result probably generalize to division rings if you put a mop somewhere, but I'm not in a position to generalize the Dual file ...

Reply:
Ah, ok, I wasn't looking carefully enough that only one was using the finite dimensionality hypothesis. By the way, I just became aware of the PerfectPairing module due to #8824. There's a `flip` specifically for maps to the dual over there.

Reply:
It's mentioned in the PR title :) It might be the case that reflexivity is a more natural setting for all these results than f.g. free modules or f.d. vector spaces, and if true we could consider another PR to change everything systematically.